### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by alexvthornton**
need to add testing

### CODE REVIEW

**Comment by Zophed** [comment]
I will review this

**Review by Zophed — COMMENTED** [review]
- There is an existing BaseLogger class. It needs an auto property to hold the class name. This property should be set in the LogFactory using an object initializer. ✔
   - Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
       - The current date/time ✔
       - The name of the class that created the logger ✔
       - The log level ✔
       - The message ✔
       - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
   - The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
   - If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
   - Inside of BaseLoggerMixins implement extension methods on BaseLogger for
       - Error, ✔
       - Warning, ✔
       - Information, and ✔
       - Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
   - Use the nameof() operator when identifying the class name to the logger ❌
   - Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
   - Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
   - Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
   - Set LangVersion to 9.0 ✔
   - Turn on Nullability (Nullable) ✔
   - All of the above should be unit tested.

- Extra Credit

   - Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌

**Comment by Zophed** [comment]
-The only test that doesn't pass tests whether the "log" method for file logger writes the information to the file; it doens't pass as the file name is hard coded into the TestMethod.
-Other item was not using the 'nameof' operator to identify the class name to the logger.
	- The code besides these two items looked good.

File: Logger.Tests/LogFactoryTests.cs [hunk 1]
```diff
@@ -6,6 +6,49 @@ namespace Logger.Tests
     [TestClass]
     public class LogFactoryTests
     {
-        
+        [TestMethod]
+        public void CreateLogger_FileLoggerNotConfigured_ReturnNull()
+        {
+            LogFactory? lf = new LogFactory();
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:12]:
```suggestion
            LogFactory? logFactory = new LogFactory();
```
I read this as `if` initially. :)

File: Logger.Tests/LogFactoryTests.cs [hunk 2]
```diff
@@ -6,6 +6,49 @@ namespace Logger.Tests
...
+
+            Assert.IsNull(baseLogger);
+        }
+
+        [TestMethod]
+        public void ConfigureFileLogger_path_StoresResultInFileLogger()
+        {
+            LogFactory? lf = new LogFactory();
+            lf.ConfigureFileLogger("Path");
+            FileLogger? fileLogger = (FileLogger?)lf.CreateLogger(nameof(LogFactoryTests));
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:23]:
```suggestion
            BaseLogger? baseLogger= lf.CreateLogger(nameof(LogFactoryTests));
```

File: Logger.Tests/LogFactoryTests.cs [hunk 3]
```diff
@@ -6,6 +6,49 @@ namespace Logger.Tests
...
+            
+            string res = "";
+            if(fileLogger != null && !string.IsNullOrEmpty(fileLogger.FilePath))
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:26]:
```suggestion

```
This if statement can be removed as `fileLogger` should not be null if it's configured - if it is the test should fail when trying to grab the string. With this in mind, you may want to write another unit test that explicitly checks if `fileLogger` is not null when configuring it's path.

File: Logger.Tests/LogFactoryTests.cs [hunk 4]
```diff
@@ -6,6 +6,49 @@ namespace Logger.Tests
...
+
+        [TestMethod]
+        public void CreateLogger_NameOfClass_StoresResultInBaseLogger()
+        {
+            LogFactory? lf = new LogFactory();
+            lf.ConfigureFileLogger("Path");
+            FileLogger? fileLogger = (FileLogger?)lf.CreateLogger(nameof(LogFactoryTests));
+            
+            string res = "";
+            if(fileLogger != null && !string.IsNullOrEmpty(fileLogger.ClassName))
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:41]:
```suggestion

```
Same suggestion as above.

File: Logger.Tests/FileLoggerTests.cs [hunk 5]
```diff
@@ -1,10 +1,62 @@
...
+            FileLogger? fileLogger = new FileLogger(null!);
+        }
+
+        [TestMethod]
+        public void FileLogger_testFilePath()
+        {
+            FileLogger? fileLogger = new FileLogger("test");
+            
+            string res = "";
+            if(fileLogger != null && !string.IsNullOrEmpty(fileLogger.FilePath))
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:25]:
```suggestion

```
See comment on line 26 of `LogFactoryTests.cs`

File: Logger/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        public FileLogger(string filePath)
+        {
+
+            FilePath = filePath;
+        }
+
+        override
+        public void Log(LogLevel logLevel, string message)
+        {
+            if(!File.Exists(FilePath))
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:29]: You may want to have this method create the file if it doesn't exist rather than throwing an exception.

File: Logger.Tests/FileLoggerTests.cs [hunk 7]
```diff
@@ -1,10 +1,62 @@
...
+            fileLogger.Log(LogLevel.Error, "message");
+            
+        }
+
+        [TestMethod]
+        public void Log_ValidFile_appendToFile()
+        {
+            LogFactory? lf = new LogFactory();
+
+            lf.ConfigureFileLogger("realFile.txt");
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:48]: If you plan to use a file in a unit test you should programmatically create and then delete that file (or have your `fileLogger` automatically make the file when `.Log` is called.)

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 8]
```diff
@@ -11,29 +11,112 @@ public class BaseLoggerMixinsTests
         [ExpectedException(typeof(ArgumentNullException))]
         public void Error_WithNullLogger_ThrowsException()
         {
-            // Arrange
+            #pragma warning disable CS8625 // Cannot convert null literal to non-nullable reference type.
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:14]:
```suggestion

```
You don't need this suppression here because the `BaseLogger` parameter is declared as a nullable value in your `BaseLoggerMixins.cs` class.

File: Logger/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            {
+                throw new FileNotFoundException();
+            }
+
+            TextWriter tw = new StreamWriter(FilePath, true);
+
+            string res = String.Format("{0} {1} {2}: {3}", DateTime.Now.ToString(), base.ClassName, logLevel, message);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:36]:
```suggestion
            string res = string.Format("{0} {1} {2}: {3}", DateTime.Now.ToString(), base.ClassName, logLevel, message);
```
You can simplify this to just use the C# keyword rather than the BCL name. You can refer to the guidelines on page 47 in chapter 2 of the book for more information about this.

**Review by mmwoodfo — COMMENTED** [review]
##  (alexvthornton)

## Assignment2

| Build/Compile (Build succeeded: 0 Error(s), 0 Warning(s))| **<10>** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |


| Assignment  | **<10 (missing lots) - 20 (minor)>** |
| --------------------- | ------ |
| <Project Assignment Goes Here> | ❌✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **❌** |
| Appropriately handle all warnings | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Logger | Enable | Logger.sln |  |  | 4 |
| Logger.Tests | Enable | Logger.sln |  |  | 4 |
| LoginStuff | Disable |  |  |  | 4 |
| LoginStuff.Tests | Disable |  |  |  | 4 |


| Testing | **<0 - 20>** |
| --------------------- | ------ |
| Test Run FAILED: Total tests=16, Passed=15, Failed=1, Skipped=0 | **87%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **<10>** |
| Most assigned work is tested or contain some errors. | **<5>** |
| All assigned work is tested | **<5>** |


| Extra Credit | **<0 - 10>** |
| --------------------- | ------ |
| <Extra Credit Goes Here> | ❌✔ |

File: Logger/LogFactory.cs [hunk 10]
```diff
@@ -2,10 +2,19 @@
...
+        private string? filePath;
+
+        public BaseLogger? CreateLogger(string className)
         {
-            
-            return null;
+            if(filePath is null)
+                return null;
+
+            return new FileLogger(filePath) {ClassName = className};
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:13]: 👍🏻
